geomstats/geometry/connection.py

```diff
@@ -41,39 +49,85 @@ def connection(self, tangent_vector_a, tangent_vector_b, base_point):
         raise NotImplementedError(
                 'connection is not implemented.')
 
-    def exp(self, tangent_vec, base_point):
+    def exp(self, tangent_vec, base_point, n_steps=N_STEPS):
         """Exponential map associated to the affine connection.
 
+        Exponential map at base_point of tangent_vec computed by integration
+        of the geodesic the equation (initial value problem), using the
+        christoffel symbols
+
         Parameters
         ----------
         tangent_vec: array-like, shape=[n_samples, dimension]
                                  or shape=[1, dimension]
 
         base_point: array-like, shape=[n_samples, dimension]
                                 or shape=[1, dimension]
-        """
-        raise NotImplementedError(
-                'The affine connection exponential is not implemented.')
 
-    def log(self, point, base_point):
+        n_steps: int
+        """
+        tangent_vec = gs.to_ndarray(tangent_vec, to_ndim=2)
+        base_point = gs.to_ndarray(base_point, to_ndim=2)
+
+        def geodesic_equation(x, v):
+            """
+            The geodesic ordinary differential equation associated
+            with the connection.
+            """
+            v = gs.to_ndarray(v, to_ndim=2)
+            gamma = self.christoffels(x)
+            return - gs.einsum('lkij,li,lj->lk', gamma, v, v)
+
+        initial_state = (base_point, tangent_vec)
+        flow, _ = self.integrate(geodesic_equation, initial_state,
+                                 n_steps=n_steps)
+        return flow[-1]
+
+    def log(self, point, base_point, n_steps=N_STEPS):
```

Comment:
🎉 woooo

Reply:
@nguigs  :carrot:  exactly what I told you in the kitchen! 
we should use '...' einsum ellipsis instead of 'to_ndarray' to vectorize consistently and return a 1D-array if given one.
+ delete line `v = gs.to_ndarray(v, to_ndim=2)`
+ replace `gs.einsum('...kij,...i,...j->...k', gamma, v, v)`
(I even think the last `->...k` can be omitted and figured out by einsum)

Reply:
We had the einsum ellipsis before, but had to replace all of them because pytorch and/or tensorflow do not support them. This might have changed as we use newer versions of pytorch and tensorflow now, but it's worth testing before converting everything.

Reply:
We had the einsum ellipsis before, but had to replace all of them because pytorch and/or tensorflow do not support them. This might have changed as we use newer versions of pytorch and tensorflow now, but it's worth testing before converting everything.

Reply:
I started changing the pytorch einsum in the backend, I will do the tests and try to get the exp and log work with this backend as well

Reply:
We need to solve #338 first and then reformat this. As it is, I think it won't work if tangent vectors are matrices.